### PR TITLE
Code gen setup functions - mainly for ABY

### PR DESCRIPTION
**Pull request by ianwilkietomasik**
This branch has all of the changes to the runtime and code generator interfaces that were necessary to support a new function `setup` for code generators, which allows for the execution of arbitrary setup code before programs are run.

### CODE REVIEW

File: runtime2/src/main/kotlin/edu/cornell/cs/apl/viaduct/runtime/Runtime.kt [hunk 1]
```diff
@@ -14,6 +16,8 @@ interface Runtime {
     suspend fun <T> receive(type: KType, sender: Host): T
 
     suspend fun <T> send(type: KType, value: T, receiver: Host)
+
+    suspend fun url(host: Host): URL
```
Comment (rolph-recto) [inline runtime2/src/main/kotlin/edu/cornell/cs/apl/viaduct/runtime/Runtime.kt:20]: why is this a URL? shouldn't it be just an IP address?

Comment (cacay) [reply runtime2/src/main/kotlin/edu/cornell/cs/apl/viaduct/runtime/Runtime.kt:20]: We need a URL for TLS. URLs can be IP if necessary.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/aby/ABYCodeGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,153 @@
+package edu.cornell.cs.apl.viaduct.backends.aby
```
Comment (cacay) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/aby/ABYCodeGenerator.kt:1]: This PR is about the infrastructure. We should save this file for the ABY PR.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/cleartext/CleartextCodeGenerator.kt [hunk 3]
```diff
@@ -415,4 +416,6 @@ class CleartextCodeGenerator(context: CodeGeneratorContext) :
         }
         return receiveBuilder.build()
     }
+
+    override fun setup(protocol: Protocol): Iterable<PropertySpec> = listOf()
```
Comment (cacay) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/cleartext/CleartextCodeGenerator.kt:420]: You should add this to `AbstractCodeGenerator` so the old code generators don't have to change. Make sure to use `open`:

```kotlin
open override fun setup(protocol: Protocol): Iterable<PropertySpec> = listOf()
```

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/CodeGeneratorContext.kt [hunk 4]
```diff
@@ -22,9 +22,13 @@ interface CodeGeneratorContext {
     // returns a fresh kotlin name for baseName
     fun newTemporary(baseName: String): String
 
+    fun tempKotlinNameToProtocol(kotlinName: String): Protocol
```
Comment (cacay) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/CodeGeneratorContext.kt:25]: What is this function for? I think we said this should be private to each `CodeGenerator` if the need it.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/Utility.kt [hunk 5]
```diff
@@ -87,3 +88,13 @@ fun receiveReplicated(
 
     return receiveExpression.build()
 }
+
+fun getRole(protocol: Protocol, host: Host): CodeBlock =
```
Comment (cacay) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/Utility.kt:92]: This is ABY specific, not a general purpose utility function. Should be moved into the `ABYCodeGenerator` class.

**Review by cacay — CHANGES_REQUESTED** [review]

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/CodeGeneratorContext.kt (hunk 4 above)
Comment (ianwilkietomasik) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/CodeGeneratorContext.kt:25]: This is also ABY specific, will remove from this PR

**Comment by ianwilkietomasik** [comment]
I removed the ABY specific stuff from this branch, so should be good now.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/cleartext/CleartextCodeGenerator.kt (hunk 3 above)
Comment (ianwilkietomasik) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/cleartext/CleartextCodeGenerator.kt:420]: I added the open keyword and IntelliJ greyed it out, wondering why?

Comment (cacay) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/cleartext/CleartextCodeGenerator.kt:420]: That's right, `overrides` are open by default. You need `open` is you declare a method in an abstract class.

**Review by cacay — APPROVED** [review]

**Pull request merged**